spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/FirestoreTemplate.java

```diff
@@ -124,34 +136,76 @@ public FirestoreTemplate(FirestoreStub firestore, String parent) {
 						.map(document -> PublicClassMapper.convertToCustomClass(document, clazz)));
 	}
 
+	public <T> Mono<Long> count(Class<T> entityClass) {
+		return findAllDocuments(entityClass, ID_PROJECTION).count();
+	}
+
 	public <T> Mono<Long> deleteAll(Class<T> clazz) {
 		return Mono.defer(() ->
-			findAllDocuments(clazz).flatMap(this::callDelete).count());
+			findAllDocuments(clazz).map(Document::getName).flatMap(this::callDelete).count());
+	}
+
+	public <T> Mono<Void> delete(Publisher<T> entityPublisher) {
+		Flux<String> idFlux = Flux.from(entityPublisher).map(this::buildResourceName);
+		return deleteAllByFullName(idFlux);
 	}
 
-	private Mono<Empty> callDelete(Document doc) {
-		DeleteDocumentRequest deleteDocumentRequest = DeleteDocumentRequest.newBuilder().setName(doc.getName())
+	public <T> Mono<Void> deleteAllById(Publisher<T> idPublisher, Class entityClass) {
+		return Mono.defer(() -> {
+			FirestorePersistentEntity<?> persistentEntity = this.mappingContext.getPersistentEntity(entityClass);
+			return Flux.from(idPublisher).flatMap(id -> callDelete(buildResourceName(persistentEntity, id.toString())))
+					.then(Mono.empty());
+		});
+	}
+
+	private Mono<Empty> callDelete(String docName) {
+		DeleteDocumentRequest deleteDocumentRequest = DeleteDocumentRequest.newBuilder().setName(docName)
 				.build();
 		return ObservableReactiveUtil.unaryCall(
 						obs -> this.firestore.deleteDocument(deleteDocumentRequest, obs));
 	}
 
+	private <T> Mono<Void> deleteAllByFullName(Publisher<T> idPublisher) {
```

Comment:
Can this be consolidated into `delete()`? that is the only caller of this helper method. Then all the public delete methods can just delegate to `callDelete`.